File: frontend/src/app/services/test-case.service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject, throwError } from 'rxjs';
import { map, catchError, shareReplay, tap } from 'rxjs/operators';
import { of } from 'rxjs';
import { API_URL } from '../app-settings';

/**
 * Field names that store multiple values (persisted as JSON arrays
 * server-side). The dropdown configuration endpoint also returns this
 * list under `multi_value_fields` — keep the two in sync if you add a
 * new column.
 */
export const TEST_CASE_MULTI_VALUE_FIELDS = [
  'reference_document',
  'associated_requirement_id',
  'screen_id',
  'feature',
  'region',
  'brand',
  'vehicle_variant',
  'vehicle_mode',
  'env_dependency',
  'testsuite_type',
] as const;

export type TestCaseMultiValueField = typeof TEST_CASE_MULTI_VALUE_FIELDS[number];

/**
 * Multi-value fields are stored server-side as JSON arrays but legacy
 * code paths (and the bulk import service) still emit plain strings. The
 * union keeps both shapes valid; consumers should normalise via the
 * `TestCaseService.mvArray` / `mvDisplay` helpers below.
 */
export type MultiValue = string | string[] | undefined;

export interface TestCase {
  id?: number;
  test_case_id: string;
  title?: string;
  description?: string;
  vehicle_model?: string;
  severity?: string;

  reference_document?: MultiValue;
  associated_requirement_id?: MultiValue;
  screen_id?: MultiValue;
  feature?: MultiValue;
  region?: MultiValue;
  brand?: MultiValue;
  vehicle_variant?: MultiValue;
  vehicle_mode?: MultiValue;
  env_dependency?: MultiValue;
  testsuite_type?: MultiValue;

  dr_applicable_screens?: string;
  dr_id?: string;
  test_objective?: string;
  preconditions?: string;
  procedure?: string;
  expected_behavior?: string;
  test_type?: string;
  vehicle_specification?: string;
  requirement_type?: string;
  /** "Yes" / "No" — single-select dropdown driven by config.yaml. */
  regulation?: string;
  priority?: string;

  created_at?: string;
  updated_at?: string;
  is_high_priority?: boolean;
  has_requirements?: boolean;
  test_complexity?: string;
}

export interface TestCaseCreateRequest {
  test_case_id: string;
  title?: string;
  description?: string;
  vehicle_model?: string;
  severity?: string;

  reference_document?: MultiValue;
  associated_requirement_id?: MultiValue;
  screen_id?: MultiValue;
  feature?: MultiValue;
  region?: MultiValue;
  brand?: MultiValue;
  vehicle_variant?: MultiValue;
  vehicle_mode?: MultiValue;
  env_dependency?: MultiValue;
  testsuite_type?: MultiValue;

  dr_applicable_screens?: string;
  dr_id?: string;
  test_objective?: string;
  preconditions?: string;
  procedure?: string;
  expected_behavior?: string;
  test_type?: string;
  vehicle_specification?: string;
  requirement_type?: string;
  regulation?: string;
  priority?: string;
}

export type TestCaseUpdateRequest = Partial<TestCaseCreateRequest>;

/**
 * Shape of GET /api/test-cases/dropdowns. Each option list is configurable
 * via `config.yaml > test_case_dropdowns`. `multi_value_fields` echoes
 * which fields the backend persists as JSON arrays so the UI can render
 * single vs. multi pickers from the same source.
 */
export interface TestCaseDropdowns {
  multi_value_fields: string[];
  feature: string[];
  test_type: string[];
  region: string[];
  brand: string[];
  vehicle_variant: string[];
  vehicle_mode: string[];
  env_dependency: string[];
  regulation: string[];
  priority: string[];
  testsuite_type: string[];
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data?: T;
  count?: number;
  error?: string;
}

/**
 * Excel import — types returned by /api/test-cases/import/preview and /import.
 */
export interface TestCaseImportSheetPreview {
  sheet: string;
  target: string | null;
  row_count_estimate: number;
  raw_headers: string[];
  /** Map of raw header → backend's auto-detected canonical field (or null). */
  suggested_mapping: { [rawHeader: string]: string | null };
  known_fields: string[];
  id_field: string | null;
  required: string[];
  sample_rows: Array<Array<string | null>>;
}

export interface TestCaseImportPreview {
  file: string;
  sheets: TestCaseImportSheetPreview[];
}

export interface TestCaseImportSheetResult {
  sheet: string;
  target: string;
  created: number;
  /** Number of existing rows updated when duplicate_strategy = 'replace'. */
  updated?: number;
  skipped: number;
  failed: number;
  errors: Array<{ row?: number; id?: string; error: string }>;
}

export interface TestCaseImportFileResult {
  file: string;
  created: number;
  updated?: number;
  skipped: number;
  failed: number;
  sheets: TestCaseImportSheetResult[];
  errors: Array<{ sheet?: string; row?: number; id?: string; error: string }>;
}

export interface TestCaseImportResult {
  files: TestCaseImportFileResult[];
  totals: { created: number; updated?: number; skipped: number; failed: number };
}

/** What to do for rows whose ID is already in the DB. */
export type TestCaseImportDuplicateStrategy = 'skip' | 'replace';

export interface TestCaseImportFieldsResponse {
  target: string;
  id_field: string;
  required: string[];
  fields: string[];
}

@Injectable({
  providedIn: 'root'
})
export class TestCaseService {
  /**
   * Coerce a multi-value field (array, CSV string, or scalar) into a
   * plain `string[]`. Empty / blank entries are dropped, so callers can
   * rely on `mvArray(...).length === 0` for the "no values" check.
   */
  static mvArray(value: MultiValue): string[] {
    if (value === null || value === undefined) {
      return [];
    }
    if (Array.isArray(value)) {
      return value
        .map(v => (v ?? '').toString().trim())
        .filter(v => v !== '');
    }
    const str = value.toString().trim();
    if (str === '') {
      return [];
    }
    if (str.startsWith('[') && str.endsWith(']')) {
      try {
        const parsed = JSON.parse(str);
        if (Array.isArray(parsed)) {
          return parsed
            .map(v => (v ?? '').toString().trim())
            .filter(v => v !== '');
        }
      } catch {
        /* fall through to CSV split */
      }
    }
    return str.split(',').map(s => s.trim()).filter(Boolean);
  }

  /** Join a multi-value field with `, ` for read-only display / search. */
  static mvDisplay(value: MultiValue, separator = ', '): string {
    return TestCaseService.mvArray(value).join(separator);
  }

  private http = inject(HttpClient);
  private readonly baseUrl = API_URL;
  
  private testCasesSubject = new BehaviorSubject<TestCase[]>([]);
  public testCases$ = this.testCasesSubject.asObservable();

  /** In-memory dropdown cache. Populated lazily by the first `getDropdowns()` call. */
  private dropdownsCache$: Observable<TestCaseDropdowns> | null = null;
  public readonly dropdowns = signal<TestCaseDropdowns | null>(null);

  constructor() {
    this.loadTestCases();
    // Warm the dropdown cache once on app boot so create / detail screens
    // never show empty selects.
    this.getDropdowns().subscribe({ error: () => { /* fail silently */ } });
  }

  /**
   * Load all test cases from the API
   */
  loadTestCases(): void {
    this.http.get<ApiResponse<TestCase[]>>(`${this.baseUrl}/test-cases/`)
      .pipe(
        map(response => response.data || []),
        catchError(() => of([]))
      )
      .subscribe(testCases => {
        this.testCasesSubject.next(testCases);
      });
  }

  /**
   * Get all test cases as Observable
   */
  getTestCases(): Observable<TestCase[]> {
    return this.http.get<ApiResponse<TestCase[]>>(`${this.baseUrl}/test-cases/`)
      .pipe(
        map(response => response.data || []),
        catchError(() => of([]))
      );
  }

  /**
   * Get test case by ID
   */
  getTestCaseById(id: string): Observable<TestCase | null> {
    return this.http.get<ApiResponse<TestCase>>(`${this.baseUrl}/test-cases/${id}`)
      .pipe(
        map(response => response.data || null),
        catchError(() => of(null))
      );
  }

  /**
   * Get test cases by feature
   */
  getTestCasesByFeature(feature: string): Observable<TestCase[]> {
    const params = new URLSearchParams();
    params.set('feature', feature);
    
    return this.http.get<ApiResponse<TestCase[]>>(`${this.baseUrl}/test-cases/feature?${params.toString()}`)
      .pipe(
        map(response => response.data || []),
        catchError(() => of([]))
      );
  }

  /**
   * Create a new test case.
   *
   * Errors must propagate to the caller — silently returning `null` on backend
   * failure causes the standalone create page to flash a "created successfully"
   * banner for rows that were actually rejected (e.g. invalid test_case_id
   * format), which is what users were seeing.
   */
  createTestCase(testCaseData: TestCaseCreateRequest): Observable<TestCase> {
    return this.http.post<ApiResponse<TestCase>>(`${this.baseUrl}/test-cases/`, testCaseData)
      .pipe(
        map(response => {
          if (response.success && response.data) {
            const currentTestCases = this.testCasesSubject.value;
            this.testCasesSubject.next([response.data, ...currentTestCases]);
            return response.data;
          }
          throw new Error(response.message || response.error || 'Failed to create test case');
        }),
        catchError(error => {
          console.error('Error creating test case:', error);
          const msg = error?.error?.message || error?.error?.error || error?.message || 'Failed to create test case';
          return throwError(() => new Error(msg));
        })
      );
  }

  /**
   * Update an existing test case
   */
  updateTestCase(id: string, testCaseData: TestCaseUpdateRequest): Observable<TestCase | null> {
    return this.http.put<ApiResponse<TestCase>>(`${this.baseUrl}/test-cases/${id}`, testCaseData)
      .pipe(
        map(response => {
          if (response.success && response.data) {
            // Update local cache
            const currentTestCases = this.testCasesSubject.value;
            const updatedTestCases = currentTestCases.map(testCase => 
              testCase.test_case_id === id ? response.data! : testCase
            );
            this.testCasesSubject.next(updatedTestCases);
            return response.data;
          }
          return null;
        }),
        catchError(error => {
          console.error('Error updating test case:', error);
          return of(null);
        })
      );
  }

  /**
   * Excel preview: parse a single workbook and return detected headers,
   * auto-mapping suggestions, and sample rows so the user can confirm the
   * mapping before importing 10k+ rows. Does not insert anything.
   */
  previewImport(file: File, sampleRows: number = 5): Observable<TestCaseImportPreview> {
    const form = new FormData();
    form.append('file', file);
    form.append('sample_rows', String(sampleRows));
    return this.http
      .post<ApiResponse<TestCaseImportPreview>>(`${this.baseUrl}/test-cases/import/preview`, form)
      .pipe(
        map(response => {
          if (!response.success || !response.data) {
            throw new Error(response.message || response.error || 'Preview failed');
          }
          return response.data;
        })
      );
  }

  /**
   * Spreadsheet bulk-import. Pass one or more `.xlsx`, `.xlsm`, or `.csv`
   * files and an optional `mapping` of {rawHeader: canonicalField}.
   * Unmapped headers fall back to auto-detection.
   *
   * `duplicateStrategy` controls what happens when an incoming row's ID is
   * already present:
   *   - `'skip'`    (default) — leave the existing row alone
   *   - `'replace'` — UPDATE the existing row with the spreadsheet values
   *
   * Returns per-sheet/per-file counts (created / updated / skipped / failed).
   */
  importTestCases(
    files: File[],
    mapping?: { [rawHeader: string]: string },
    duplicateStrategy: TestCaseImportDuplicateStrategy = 'skip',
  ): Observable<TestCaseImportResult> {
    const form = new FormData();
    files.forEach(f => form.append('files', f));
    if (mapping && Object.keys(mapping).length > 0) {
      form.append('mapping', JSON.stringify(mapping));
    }
    form.append('duplicate_strategy', duplicateStrategy);
    return this.http
      .post<ApiResponse<TestCaseImportResult>>(`${this.baseUrl}/test-cases/import`, form)
      .pipe(
        map(response => {
          if (!response.data) {
            throw new Error(response.message || response.error || 'Bulk import failed');
          }
          // Refresh local cache after a successful import so the table
          // immediately reflects newly created OR updated rows.
          const t = response.data.totals;
          if ((t?.created || 0) > 0 || (t?.updated || 0) > 0) {
            this.loadTestCases();
          }
          return response.data;
        })
      );
  }

  /**
   * Fetch dropdown / multi-select options sourced from `config.yaml`.
   * Cached for the lifetime of the service so create + detail screens
   * share a single network call.
   */
  getDropdowns(forceReload = false): Observable<TestCaseDropdowns> {
    if (forceReload) {
      this.dropdownsCache$ = null;
    }
    if (!this.dropdownsCache$) {
      this.dropdownsCache$ = this.http
        .get<ApiResponse<TestCaseDropdowns>>(`${this.baseUrl}/test-cases/dropdowns`)
        .pipe(
          map(response => {
            if (!response.success || !response.data) {
              throw new Error(response.message || response.error || 'Failed to load dropdowns');
            }
            return response.data;
          }),
          tap(data => this.dropdowns.set(data)),
          shareReplay({ bufferSize: 1, refCount: false }),
        );
    }
    return this.dropdownsCache$;
  }

  /** Canonical fields for the test_cases target — used by the mapping UI dropdowns. */
  getImportFields(): Observable<TestCaseImportFieldsResponse> {
    return this.http
      .get<ApiResponse<TestCaseImportFieldsResponse>>(`${this.baseUrl}/test-cases/import/fields`)
      .pipe(
        map(response => {
          if (!response.success || !response.data) {
            throw new Error(response.message || response.error || 'Failed to load fields');
          }
          return response.data;
        })
      );
  }

  /**
   * Delete a test case
   */
  deleteTestCase(id: string): Observable<boolean> {
    return this.http.delete<ApiResponse<any>>(`${this.baseUrl}/test-cases/${id}`)
      .pipe(
        map(response => {
          if (response.success) {
            // Update local cache
            const currentTestCases = this.testCasesSubject.value;
            const filteredTestCases = currentTestCases.filter(testCase => testCase.test_case_id !== id);
            this.testCasesSubject.next(filteredTestCases);
            return true;
          }
          return false;
        }),
        catchError(() => of(false))
      );
  }

  /**
   * Get current test cases from subject
   */
  getCurrentTestCases(): TestCase[] {
    return this.testCasesSubject.value;
  }

  /**
   * Validate test case ID format (XX_FEATURE_XXX1)
   */
  isValidTestCaseId(testCaseId: string): boolean {
    const pattern = /^[A-Z]{2}_[A-Z_]+_\d+$/;
    return pattern.test(testCaseId);
  }

  /**
   * Validate priority against the configured dropdown values, falling
   * back to the historical hard-coded list if config has not loaded yet.
   */
  isValidPriority(priority: string): boolean {
    const configured = this.dropdowns()?.priority;
    return (configured && configured.length ? configured : ['P1', 'P2', 'P3', 'P4']).includes(priority);
  }

  isValidTestType(testType: string): boolean {
    const configured = this.dropdowns()?.test_type;
    return (configured && configured.length ? configured : ['Positive', 'Negative', 'Abnormal', 'Boundary']).includes(testType);
  }

  /**
   * Check if test case ID is available
   */
  isTestCaseIdAvailable(testCaseId: string, currentTestCaseId?: string): boolean {
    const currentTestCases = this.testCasesSubject.value;
    return !currentTestCases.some(testCase => 
      testCase.test_case_id === testCaseId && 
      testCase.test_case_id !== currentTestCaseId
    );
  }

  /**
   * Get unique features from current test cases
   */
  getUniqueFeatures(): string[] {
    // `feature` is multi-value (string | string[]); flatten via mvArray then dedupe.
    const features = this.testCasesSubject.value
      .flatMap(tc => TestCaseService.mvArray(tc.feature));
    return [...new Set(features)].sort();
  }

  /**
   * Get unique priorities from current test cases
   */
  getUniquePriorities(): string[] {
    return this.dropdowns()?.priority ?? ['P1', 'P2', 'P3', 'P4'];
  }

  getUniqueTestTypes(): string[] {
    return this.dropdowns()?.test_type ?? ['Positive', 'Negative', 'Abnormal', 'Boundary'];
  }
}

File: backend/src/services/test_case_service.py
from abc import ABC, abstractmethod
from typing import List, Dict, Any, Optional
from src.schemas.test_case_schema import TestCaseSchema, TestCaseCreateSchema
from src.repositories.test_case_repository import ITestCaseRepository
from src.infrastructure.configuration_manager import get_config_manager
from src.infrastructure.logging_config import get_logger

logger = get_logger(__name__)


# Sensible defaults if config.yaml is missing the dropdown block. Validation
# should never block writes when config is unavailable, just fall back to
# the historical hard-coded list.
_DEFAULT_PRIORITIES = ["P1", "P2", "P3", "P4"]
_DEFAULT_TEST_TYPES = ["Positive", "Negative", "Abnormal", "Boundary"]
_DEFAULT_REGULATIONS = ["Yes", "No"]


def _allowed_values(key: str, fallback: List[str]) -> List[str]:
    """Pull ``test_case_dropdowns.<key>`` out of config, falling back if absent."""
    try:
        dropdowns = get_config_manager().get_test_case_dropdowns()
    except Exception:
        return fallback
    values = dropdowns.get(key) or []
    return [str(v) for v in values] if values else fallback


class ITestCaseService(ABC):
    """Interface for test case business logic operations"""
    
    @abstractmethod
    def get_all_test_cases(self) -> List[Dict[str, Any]]:
        """Get all test cases with business logic"""
        pass
    
    @abstractmethod
    def get_test_case_by_id(self, test_case_id: str) -> Optional[Dict[str, Any]]:
        """Get test case by ID with business logic"""
        pass
    
    @abstractmethod
    def get_test_cases_by_feature(self, feature: str) -> List[Dict[str, Any]]:
        """Get test cases by feature with business logic"""
        pass
    
    @abstractmethod
    def create_test_case(self, test_case_data: TestCaseCreateSchema) -> Dict[str, Any]:
        """Create test case with business logic validation"""
        pass
    
    @abstractmethod
    def update_test_case(self, test_case_id: str, test_case_data: dict) -> Optional[Dict[str, Any]]:
        """Update test case with business logic validation"""
        pass
    
    @abstractmethod
    def delete_test_case(self, test_case_id: str) -> bool:
        """Delete test case with business logic validation"""
        pass


class TestCaseService(ITestCaseService):
    """Concrete implementation of test case service with business logic"""
    
    def __init__(self, test_case_repository: ITestCaseRepository):
        self.test_case_repository = test_case_repository
    
    def get_all_test_cases(self) -> List[Dict[str, Any]]:
        """Get all test cases with business logic"""
        try:
            test_cases = self.test_case_repository.find_all()
            
            # Apply business logic transformations
            for test_case in test_cases:
                # Add computed fields
                test_case['is_high_priority'] = test_case.get('priority') == 'P1'
                test_case['has_requirements'] = bool(test_case.get('associated_requirement_id'))
                test_case['test_complexity'] = self._calculate_test_complexity(test_case)
            
            return test_cases
        except Exception as e:
            raise Exception(f"Service error: Failed to get test cases - {str(e)}")
    
    def get_test_case_by_id(self, test_case_id: str) -> Optional[Dict[str, Any]]:
        """Get test case by ID with business logic"""
        try:
            if not test_case_id or not isinstance(test_case_id, str):
                raise ValueError("Invalid test case ID")
            
            test_case = self.test_case_repository.find_by_id(test_case_id)
            if test_case:
                # Apply business logic transformations
                test_case['is_high_priority'] = test_case.get('priority') == 'P1'
                test_case['has_requirements'] = bool(test_case.get('associated_requirement_id'))
                test_case['test_complexity'] = self._calculate_test_complexity(test_case)
            
            return test_case
        except Exception as e:
            raise Exception(f"Service error: Failed to get test case {test_case_id} - {str(e)}")
    
    def get_test_cases_by_feature(self, feature: str) -> List[Dict[str, Any]]:
        """Get test cases by feature with business logic"""
        try:
            if not feature or not isinstance(feature, str):
                raise ValueError("Invalid feature name")
            
            test_cases = self.test_case_repository.find_by_feature(feature)
            
            # Apply business logic transformations
            for test_case in test_cases:
                test_case['is_high_priority'] = test_case.get('priority') == 'P1'
                test_case['has_requirements'] = bool(test_case.get('associated_requirement_id'))
                test_case['test_complexity'] = self._calculate_test_complexity(test_case)
            
            return test_cases
        except Exception as e:
            raise Exception(f"Service error: Failed to get test cases for feature {feature} - {str(e)}")
    
    def create_test_case(self, test_case_data: TestCaseCreateSchema) -> Dict[str, Any]:
        """Create test case with business logic validation"""
        # Validation errors must propagate as ValueError so the controller can map
        # them to HTTP 400 instead of the generic 500. Wrapping every exception
        # in `Exception(...)` previously hid the type and caused the frontend to
        # see a 500, swallow it as null, and falsely report "created successfully".
        try:
            self._validate_test_case_creation(test_case_data)

            existing_test_case = self.test_case_repository.find_by_id(test_case_data.test_case_id)
            if existing_test_case:
                raise ValueError(f"Test case ID '{test_case_data.test_case_id}' already exists")

            test_case = self.test_case_repository.create(test_case_data)

            if test_case:
                test_case['is_high_priority'] = test_case.get('priority') == 'P1'
                test_case['has_requirements'] = bool(test_case.get('associated_requirement_id'))
                test_case['test_complexity'] = self._calculate_test_complexity(test_case)

            return test_case
        except ValueError:
            raise
        except Exception as e:
            raise Exception(f"Service error: Failed to create test case - {str(e)}")
    
    def update_test_case(self, test_case_id: str, test_case_data: dict) -> Optional[Dict[str, Any]]:
        """Update test case with business logic validation"""
        try:
            if not test_case_id or not isinstance(test_case_id, str):
                raise ValueError("Invalid test case ID")
            
            # Check if test case exists
            existing_test_case = self.test_case_repository.find_by_id(test_case_id)
            if not existing_test_case:
                raise ValueError(f"Test case with ID {test_case_id} not found")
            
            # Business logic validation
            self._validate_test_case_update(test_case_data)
            
            # Update test case
            test_case = self.test_case_repository.update(test_case_id, test_case_data)
            
            # Apply business logic transformations
            if test_case:
                test_case['is_high_priority'] = test_case.get('priority') == 'P1'
                test_case['has_requirements'] = bool(test_case.get('associated_requirement_id'))
                test_case['test_complexity'] = self._calculate_test_complexity(test_case)

            return test_case
        except Exception as e:
            raise Exception(f"Service error: Failed to update test case {test_case_id} - {str(e)}")
    
    def delete_test_case(self, test_case_id: str) -> bool:
        """Delete test case with business logic validation"""
        try:
            if not test_case_id or not isinstance(test_case_id, str):
                raise ValueError("Invalid test case ID")
            
            # Check if test case exists
            existing_test_case = self.test_case_repository.find_by_id(test_case_id)
            if not existing_test_case:
                raise ValueError(f"Test case with ID {test_case_id} not found")
            
            # Business logic: Check if test case is referenced by other entities
            # This could be expanded to check for dependencies
            
            success = self.test_case_repository.delete(test_case_id)
            return success
        except Exception as e:
            raise Exception(f"Service error: Failed to delete test case {test_case_id} - {str(e)}")
    
    def _validate_test_case_creation(self, test_case_data: TestCaseCreateSchema) -> None:
        """Validate test case creation business rules.

        Allowed values for ``priority``, ``test_type`` and ``regulation``
        are pulled from ``config.yaml > test_case_dropdowns`` so QA leads
        can extend the option lists without code changes.
        """
        if not self._is_valid_test_case_id_format(test_case_data.test_case_id):
            raise ValueError(
                "Test case ID must start with uppercase letters and end with _<digits> "
                "(e.g. TC_0001 or TC_FEATURE_001)"
            )

        priorities = _allowed_values("priority", _DEFAULT_PRIORITIES)
        test_types = _allowed_values("test_type", _DEFAULT_TEST_TYPES)
        regulations = _allowed_values("regulation", _DEFAULT_REGULATIONS)

        if test_case_data.priority and test_case_data.priority not in priorities:
            raise ValueError(f"Priority must be one of: {', '.join(priorities)}")

        if test_case_data.test_type and test_case_data.test_type not in test_types:
            raise ValueError(f"Test type must be one of: {', '.join(test_types)}")

        if test_case_data.regulation and test_case_data.regulation not in regulations:
            raise ValueError(f"Regulation must be one of: {', '.join(regulations)}")

    def _validate_test_case_update(self, test_case_data: dict) -> None:
        """Validate test case update business rules — config-driven."""
        priorities = _allowed_values("priority", _DEFAULT_PRIORITIES)
        test_types = _allowed_values("test_type", _DEFAULT_TEST_TYPES)
        regulations = _allowed_values("regulation", _DEFAULT_REGULATIONS)

        if "priority" in test_case_data and test_case_data["priority"] and test_case_data["priority"] not in priorities:
            raise ValueError(f"Priority must be one of: {', '.join(priorities)}")

        if "test_type" in test_case_data and test_case_data["test_type"] and test_case_data["test_type"] not in test_types:
            raise ValueError(f"Test type must be one of: {', '.join(test_types)}")

        if "regulation" in test_case_data and test_case_data["regulation"] and test_case_data["regulation"] not in regulations:
            raise ValueError(f"Regulation must be one of: {', '.join(regulations)}")
    
    def _is_valid_test_case_id_format(self, test_case_id: str) -> bool:
        """Validate test case ID format.

        Accepts both the short auto-generated form (``TC_0001``) and the
        legacy feature-segmented form (``TC_AAOS_BT_001``). The id must:
          * start with two or more uppercase letters
          * optionally include any number of ``_LETTER_OR_UNDERSCORE`` segments
          * end with ``_<digits>``
        """
        if not test_case_id:
            return False
        import re
        pattern = r'^[A-Z]{2,}(?:_[A-Z_]+)*_\d+$'
        return bool(re.match(pattern, test_case_id))
    
    def _calculate_test_complexity(self, test_case: Dict[str, Any]) -> str:
        """Calculate test complexity based on various factors"""
        complexity_score = 0
        
        # Factor in procedure length
        procedure = test_case.get('procedure') or ''
        if len(procedure) > 1000:
            complexity_score += 2
        elif len(procedure) > 500:
            complexity_score += 1
        
        # Factor in preconditions
        preconditions = test_case.get('preconditions') or ''
        if len(preconditions) > 500:
            complexity_score += 1
        
        # Factor in expected behavior
        expected_behavior = test_case.get('expected_behavior') or ''
        if len(expected_behavior) > 500:
            complexity_score += 1
        
        # Factor in priority
        priority = test_case.get('priority') or ''
        if priority == 'P1':
            complexity_score += 1
        
        # Determine complexity level
        if complexity_score >= 4:
            return 'High'
        elif complexity_score >= 2:
            return 'Medium'
        else:
            return 'Low'
